paychmgr/simple.go

```diff
@@ -17,7 +22,211 @@ import (
 	"github.com/filecoin-project/lotus/chain/types"
 )
 
-func (pm *Manager) createPaych(ctx context.Context, from, to address.Address, amt types.BigInt) (cid.Cid, error) {
+type paychApi interface {
+	StateWaitMsg(ctx context.Context, msg cid.Cid, confidence uint64) (*api.MsgLookup, error)
+	MpoolPushMessage(ctx context.Context, msg *types.Message) (*types.SignedMessage, error)
+}
+
+// paychFundsRes is the response to a create channel or add funds request
+type paychFundsRes struct {
+	channel address.Address
+	mcid    cid.Cid
+	err     error
+}
+
+type onCompleteFn func(*paychFundsRes)
+
+// fundsReq is a request to create a channel or add funds to a channel
+type fundsReq struct {
+	ctx        context.Context
+	from       address.Address
+	to         address.Address
+	amt        types.BigInt
+	onComplete onCompleteFn
+}
+
+// getPaych ensures that a channel exists between the from and to addresses,
+// and adds the given amount of funds.
+// If the channel does not exist a create channel message is sent and the
+// message CID is returned.
+// If the channel does exist an add funds message is sent and both the channel
+// address and message CID are returned.
+// If there is an in progress operation (create channel / add funds), getPaych
+// blocks until the previous operation completes, then returns both the channel
+// address and the CID of the new add funds message.
+// If an operation returns an error, subsequent waiting operations will still
+// be attempted.
+func (ca *channelAccessor) getPaych(ctx context.Context, from, to address.Address, amt types.BigInt) (address.Address, cid.Cid, error) {
+	// Add the request to add funds to a queue and wait for the result
+	promise := ca.enqueue(&fundsReq{ctx: ctx, from: from, to: to, amt: amt})
+	select {
+	case res := <-promise:
+		return res.channel, res.mcid, res.err
+	case <-ctx.Done():
+		return address.Undef, cid.Undef, ctx.Err()
+	}
+}
+
+// Queue up an add funds operation
+func (ca *channelAccessor) enqueue(task *fundsReq) chan *paychFundsRes {
+	promise := make(chan *paychFundsRes)
+	task.onComplete = func(res *paychFundsRes) {
+		select {
+		case <-task.ctx.Done():
+		case promise <- res:
+		}
+	}
+
+	ca.lk.Lock()
+	defer ca.lk.Unlock()
+
+	ca.fundsReqQueue = append(ca.fundsReqQueue, task)
+	go ca.processNextQueueItem()
+
+	return promise
+}
+
+// Run the operation at the head of the queue
+func (ca *channelAccessor) processNextQueueItem() {
+	ca.lk.Lock()
+	defer ca.lk.Unlock()
+
+	if len(ca.fundsReqQueue) == 0 {
+		return
+	}
+
+	head := ca.fundsReqQueue[0]
+	res := ca.processTask(head.ctx, head.from, head.to, head.amt, head.onComplete)
```

Comment:
I'm pretty sure this was already discussed - but it feels like it should be fairly easy to 'merge' multiple tasks together, so if we need to add funds to 10 different lanes, we'd still only have to block once.

(If this is more than ~15min of work, probably do as a followup PR)

Reply:
I agree that should be easy to do - I suggest doing it in a follow up PR so we can get this one in today